analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Constraints/AssignmentOrReferenceParameterConstraint.cs

```diff
@@ -0,0 +1,32 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.SymbolicExecution.Constraints;
+
+internal sealed class AssignmentOrReferenceParameterConstraint : SymbolicConstraint
+{
+    public static readonly AssignmentOrReferenceParameterConstraint AssignedOrPassedAsReference = new(ConstraintKind.AssignedOrPassedAsReference);
```

Comment:
Educational: my understanding is that this constraint captures two scenarios:
1. the parameter is assigned
2. the parameter is passed as a reference (which I guess it's only `ref`, and not `out` - because `out` have to be assigned in the method itself)

Would it make sense to have two different `AssignmentOrReferenceParameterConstraint` instances, one for each? YAGNI-wise, it's not necessarily more code, but the real question here is: are those constraints conceptually tightly coupled, in a way that having a `ThisOrThat...` constraint better than having two different constraints?

@martin-strecker-sonarsource I would like to have your thoughts on the subjects, too.

Reply:
Thanks both for your explanations!